implement ssh connector

garlick opened the pull request:
As discussed in #644, this PR implements a new ssh:// connector.

Unfinished but wanted to get it up here for any prelim. review.

I had to create a new `popen()` like function called `pxopen()` that uses a socketpair to establish bidirectional communication with a subprocess, which will usually be /usr/bin./ssh or maybe /usr/bin/rsh on our clusters.  It seems a bit odd to have `flux_open()` behind the scenes do a fork/exec - hopefully that is not going to cause problems.

The remote command is currently `flux proxy` as proposed in #645 although maybe the first pass on this should just use netcat.

garlick commented:
Hmm, the gcc+install build failed here
```
ok 20 - wreckrun: oversubscription of tasks
ok 21 - wreckrun: uneven distribution with -n, -N
ok 22 - wreckrun: too many nodes requested fails
ok 23 - wreckrun: no nnodes or ntasks args runs one task on rank 0
No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.
```
so I restarted it.

garlick commented:
Rebased on master, set O_CLOEXEC on parent fd, and added missing close of child fd after dup.

grondo commented:
This is pretty cool.

When looking this over casually I kept getting confused on the difference between the ssh connector and a `flux-proxy` type command. Is the `ssh` connector only meant to be used by `flux proxy`? Or does the ssh connector *invoke* flux-proxy?

If ssh connector use case is on a user's desktop to connect to a job running on a cluster, I worry a little about the fork/exec per `flux_open()` since there could potentially be many `flux_open` calls if a user is running a series of utilities via a shell script, etc. Perhaps I'm making too much of it now before we've even seen a problem though.

However, perhaps we could consider an approach where the connector ssh channel could be created once and be reused by multiple `flux_open()` calls? This would work easiest if you could forward the domain socket as we discussed in #644. In fact the interface might mimic `flux start` in that you do something like: 
```
flux remote-connect [OPTIONS] <remote-uri> [command]
```

Where `command` is `$SHELL` if not specified.

`remote-connect` would set up the ssh tunnel, set `FLUX_URI` appropriately, and start command(s). Once command exits the ssh tunnel is destroyed.

I hesitate to mention this, but it would be really interesting if `remote-connect` actually started a broker, then `remote-connect` could just be a special case of `start`?? :astonished: (`flux start --remote-connect=URI ...`). A broker spliced into an instance from the user's desktop might open some interesting possibilities...

Apologies if I am getting WAY off topic.

grondo commented:
This is minor, but typical name for bidirectional `popen(3)` is `popen2()` e.g. in python, or Perl has `open2` and `open3` (with stderr). If you switched the connector to using a flux reactor I wonder if you could just use libsubprocess (though that might be more trouble than it is worth).
Sorry for the "monday morning quarterbacking"

garlick commented:
> Is the ssh connector only meant to be used by flux proxy? Or does the ssh connector invoke flux-proxy?

I was thinking both.  Here are some use cases to help explain:

Case 1:  User on login node wants to run `flux-up` on job whose rank 0 is on compute node hype222
```
$ export FLUX_SSH=/usr/bin/rsh  # munged rsh for LLNL
$ flux --uri ssh://hype222/tmp/garlick/flux-123456-42/0 up
```
The remote command executed on hype222 would be
```
flux --uri local:///tmp/garlick/flux-123456-42/0 proxy --stdio
```

Case 2: User on login node wants to run SCRIPT in environment with FLUX_URI pointing to a local socket that is forwarded to the remote one.
```
flux --uri ssh://hype222/tmp/garlick/flux-123456-42/0 proxy SCRIPT
```
Again the remote command would be as above, so there are two flux-proxy invocations involved here.

Case 3: Same as case 2 except user is on desktop and we've solved the service discovery debacle:
```
flux --uri ssh://hype.llnl.gov?jobid=carrot-hottub-jupiter proxy SCRIPT
```
Remote command would be something like
```
flux proxy --jobid=carrot-hottub-jupiter --stdin
```
and then flux-proxy would use TBD magic to map jobid to URI, which could be `ssh://` type to reach compute node from login node.   (Obviously then case 1 and 2 could use `ssh://localhost&jobid=` instead.

Is this getting too weird?

garlick commented:
> A broker spliced into an instance from the user's desktop might open some interesting possibilities...

Yeah, like a local content cache!  Hmm....

garlick commented:
> This is minor, but typical name for bidirectional popen(3) is popen2() e.g. in python, or Perl has open2 and open3 (with stderr)

We can change the name if you think that would be more clear.  I was unaware of those implementations, despite quite a bit of searching around, which is a little bit disturbing :-/

> If you switched the connector to using a flux reactor I wonder if you could just use libsubprocess (though that might be more trouble than it is worth).

I started with libsubprocess but since a `flux_t` handle needs to work in reactor and non-reactor environments, and libsubprocess is hardwired to provide its own reactor-based buffering in front of the subcommand's stdio fd's, I abandoned that approach and just went with this one.

grondo commented:
I'm sorry, this should probably be easy to understand but my brain doesn't seem to be working this AM.

I  think `flux proxy` is exactly what I was thinking of as `remote-connect` above, sorry I missed that at first. Since the "remote URI" is basically required, would it make more sense for that to be an argument to `flux-proxy` and not the `flux` cmddriver itself?

e.g.
```
flux proxy from-uri [to]
```
Where `to` is by default a new shell, but could be `-` for `stdin` or a COMMAND to run.

I admit I might be totally missing what you're going for here. I apologize

grondo commented:
> I started with libsubprocess but since a flux_t handle needs to work in reactor and non-reactor environments, and libsubprocess is hardwired to provide its own reactor-based buffering in front of the subcommand's stdio fd's, I abandoned that approach and just went with this one.

Ah that is disappointing. It is sad to see direct use of `poll` when we have libev embedded in the project... I'm sorry about the subprocess stuff it obviously doesn't work as intended.

garlick commented:
> Ah that is disappointing. It is sad to see direct use of poll when we have libev embedded in the project... I'm sorry about the subprocess stuff it obviously doesn't work as intended.

The connector requirements are a bit odd, e.g. see  pollfd and pollevents callbacks.   No worries.

grondo commented:
> Ah that is disappointing. It is sad to see direct use of poll when we have libev embedded in the project.

I reread this and I want to clarify I wasn't complaining about the code in this PR, which seems nice and clean and I would not want to unnecessarily complicate it. Sorry for the confusion!

garlick commented:
> Since the "remote URI" is basically required, would it make more sense for that to be an argument to flux-proxy and not the flux cmddriver itself?

It's only required if FLUX_URI isn't set, like all flux commands.  I'm not sure I see the need to single it out; though I _do_ see that its a bit awkward having the command driver take the `--uri` argument in the general case.

My first attempts at command line designs are notoriously bad IMO so I will keep an open mind if you want to keep pushing that one :-)

garlick commented:
> I reread this and I want to clarify I wasn't complaining about the code in this PR

I didn't think you were, since those weird callbacks in connector implementations arise from the integration of flux_t handles with libev.

I've made excellent use of libsubprocess in the past - `flux-start` worked out really nicely for example.  No complaints that it didn't plug in to this case which needs to be pretty bare metal.

garlick commented:
> Since the "remote URI" is basically required, would it make more sense for that to be an argument to flux-proxy and not the flux cmddriver itself?

Were you suggesting we drop the `--uri` argument from the command driver?  That would short-circuit my argument for consistency.  The `--uri` argument is a pain to use, and I usually end up setting FLUX_URI in a sub-shell if I'm poking at a job from outside.

If `flux-proxy` had a `--uri` argument and perhaps some smarter options such as `--jobid` to derive URI's e.g. #650), it could provide that sub-shell, even skipping the proxied message routing if the job is local.

grondo commented:
> Were you suggesting we drop the --uri argument from the command driver? That would short-circuit my argument for consistency. The --uri argument is a pain to use, and I usually end up setting FLUX_URI in a sub-shell if I'm poking at a job from outside.

I was merely making a point that from a usability standpoint, the "remote URI" argument for `flux-proxy` should be an argument to the `proxy` subcommand, not use the existing `--uri` argument on the `flux` cmddriver.

I think intuitively that will make a lot more sense, and perhaps one reason why is that we are talking about different *kinds* of URIs here. Yes, they are both URIs to Flux instances, but the `flux --uri`  is a cmdline option to override or explicitly set the URI of your *enclosing instance* for a subcommand, where as the URI argument to `flux proxy` is a "remote URI" to which you want to proxy a local connector. (I could still be misunderstanding flux-proxy though, so bear with me if so). It doesn't make logical sense (to me) to be proxying your enclosing instance URI does it? (I'm thinking aloud here)

Now that you mention it though, I do wonder if we could just drop `--uri` from `flux(1)`. I can't think of any use cases where it would be needed over `FLUX_URI=... flux`, but again I could be missing something.

garlick commented:
I went ahead and added `--uri` to `flux-proxy` in this latest push.  Just to be sure we're clear, there's no functional difference between these three commands:
```
flux --uri URI proxy CMD
flux proxy --uri URI CMD
FLUX_URI=URI flux proxy CMD
```
In all cases URI is the actual URI of the instance being connected to.  FLUX_URI in the environment of CMD will be a different URI created and owned by the proxy command.  If you ran

grondo commented:
> flux --uri URI proxy CMD
> flux proxy --uri URI CMD
> FLUX_URI=URI flux proxy CMD

So, inside a flux instance, running `flux proxy CMD` would run CMD under a proxy to the current instance?

I feel like I'm definitely missing something. Would you ever want to use flux-proxy on the current instance? If not, then I'd say make `URI` a mandatory argument

```
flux proxy URI CMD
``` 

It should probably be an error if URI == FLUX_URI, because that might mean the user made a mistake? Sorry I'm not getting it...

garlick commented:
> Would you ever want to use flux-proxy on the current instance? If not, then I'd say make URI a mandatory argument

Ah understood!  That use didn't raise any red flags for me for some reason, but you're probably right - it's useless and I could see people wondering why it should work.  Will fix.

garlick commented:
With this last update, I think this PR is more or less feature complete in that `flux-proxy` now spawns a command (or interactive shell),  is cleaning up properly, and the sharness test coverage is improved.   Still needs:
* Test from hype login node to slurm-launched instance (may need to set FLUX_SSH to /usr/bin/mrsh)
* Test from desktop to instance running on hype login node (test ssh prompting for passwd, etc)
* See how proxy holds up when ssh connection is high latency (e.g. desktop to hype login)
* Man page flux-proxy(1)

I'll try to knock this off tomorrow AM.  Meanwhile, any feedback, especially on the user interface, would be much appreciated.

garlick commented:
Oh I just remembered I never did change pxopen/pxclose to popen2/pclose2.  Will do that as well.

grondo commented:
I'll try to test it out tomorrow. A good test for the UI is if I can figure out how to use it easily with just the manpage ;-)

garlick commented:
@grondo FYI I'm running into some problems on hype.  With flux installed to $HOME/local and .bashrc containing
```
PATH=$HOME/local/bin:$PATH
export LD_LIBRARY_PATH=$HOME/local/lib:$PATH
export TMPDIR=/var/tmp/garlick
```
I can start a Flux job under slurm (rank 0 on hype335, jobid 1147790.0) and this works from a login node
```
$ rsh hype335 flux proxy --jobid=1147790.0 flux up
ok:     [0-15]
slow:   
fail:   
unknown:
$
```
however this doesn't
```
$ flux proxy --uri=ssh://hype335?jobid=1147790.0 flux up
flux-proxy: ssh://hype335?jobid=1147790.0: Connection reset by peer
```
I had different problems starting a standalone job on hype login, and connecting to it from my desktop:
```
$ ./flux -t proxy --uri 'ssh://hype?jobid=31715'
--------------------------------------
>[000] ||
>[011] cmb.attrget
>[018] { "name": "rank" }
>[012] 8E01010FFFFFFFFF00000001
--------------------------------------
<[011] cmb.attrget
<[028] { "value": "0", "flags": 5 }
<[012] 8E0102070000000000000001
$ flux up
--------------------------------------
>[000] ||
>[007] kvs.get
>[028] { "conf.live.status": null }
>[012] 8E01010FFFFFFFFF00000001
--------------------------------------
>[032] |D953A|
>[007] kvs.get
>[028] { "conf.live.status": null }
>[012] 8E01010FFFFFFFFF00000001
```
Anyway more work to do here.

grondo commented:
> however this doesn't

> $ flux proxy --uri=ssh://hype335?jobid=1147790.0 flux up
> flux-proxy: ssh://hype335?jobid=1147790.0: Connection reset by peer

Is that with `FLUX_SSH=mrsh`?

garlick commented:
I change the default to /usr/bin/rsh since on our systems that's a link to mrsh and on debian systems, its a link to whatever alternatives sets it to (ssh on mine).  Seemed like slightly better odds for success on that one than ssh...

grondo commented:
Ok, thanks for clarification. Just wondering how I could start experimenting. My first very minor observation is whether we could get rid of the `--uri | --jobid` options to simplify `flux-proxy` interface even further? 

Actually, URI could be required and we could designate a URI "scheme" for Flux jobids?
This simplifies the code (you just have to check for one valid arg instead of the usual checks for one-and-one-only-option), and IMO improves the interface vastly (one of my personal peeves is "required" options). Finally, since the connection method is specified in the scheme, this is much more easily extended than adding more `--<type>=ARG` options to the command.

You could specify jobids as `flux proxy job://123456` or even say that without a `scheme://` a job target is assumed `flux proxy 123456`, though I like the former better.

Sorry I know this is really minor but will be a usability improvement IMO. As always completely open to keeping it how it is as well. Just wanted to make my case.

grondo commented:
> I change the default to /usr/bin/rsh since on our systems that's a link to mrsh and on debian systems, its a link to whatever alternatives sets it to 

Should the scheme be `rsh://` then? Might be confusing to people that don't have /usr/bin/rsh or have both and expect `ssh://` to use /usr/bin/ssh. You could probably support both easily. I don't feel too strongly either way.

grondo commented:
You probably already got this, but I found that making the following change got a little further:
```diff
diff --git a/src/connectors/ssh/ssh.c b/src/connectors/ssh/ssh.c
index 465ef7e..5b5f163 100644
--- a/src/connectors/ssh/ssh.c
+++ b/src/connectors/ssh/ssh.c
@@ -333,7 +333,7 @@ flux_t connector_init (const char *path, int flags)
 
     if (!(c->ssh_cmd = getenv ("FLUX_SSH")))
         c->ssh_cmd = default_ssh_cmd;
-    if (argz_add (&c->ssh_argz, &c->ssh_argz_len, "ssh") != 0)
+    if (argz_add (&c->ssh_argz, &c->ssh_argz_len, c->ssh_cmd) != 0)
         goto error;
     if (parse_ssh_port (c, path) < 0) /* [-p port] */
         goto error;
```

Now the problem is just that we need to use the full path to `flux` binary?
```
$ src/cmd/flux proxy --uri=ssh://hype347?jobid=114779.0
bash: flux: command not found
flux-proxy: ssh://hype347?jobid=114779.0: Protocol error
```

garlick commented:
> Just wondering how I could start experimenting.

`flux-proxy --uri=$FLUX_URI` seems to work at any depth so it seems the message routing is OK in the proxy command.   I must have botched something in the ssh:// connector.

> You probably already got this, but I found that making the following change got a little further:

Whoops!

> Now the problem is just that we need to use the full path to flux binary?

See my comments earlier about setting TMPDIR, PATH, and LD_LIBRARY_PATH in .bashrc.

grondo commented:
> See my comments earlier about setting TMPDIR, PATH, and LD_LIBRARY_PATH in .bashrc.

Ah, sorry for missing that!
How is that going to work for flux-core installed in `/opt`? Should ssh connector exec a fully qualified path to flux, determined from the calling side by default but overridden by an environment variable?

grondo commented:
Within my flux instance, this works nicely!
```
 grondo@hype347:~/git/f$ flux proxy --jobid=1147793.0 flux up
ok:     [0-3]
slow:   
fail:   
unknown:
```

However, from hype login node I get:
```
$ src/cmd/flux proxy --uri=ssh://hype347?jobid=1147793.0
flux-proxy: could not find job 1147793.0
flux-proxy: ssh://hype347?jobid=1147793.0: Protocol error
```

Am I specifying remote incorrectly?

garlick commented:
Try setting TMPDIR for the remote command.

garlick commented:
Apparently I've introduced an intermittent travis failure:
```
not ok 14 - write to fd worked
#   Failed test 'write to fd worked'
#   at ../../../../src/common/libutil/test/popen2.c line 59.
ok 15 - pclose2 OK
ok 16 - popen2 /bin/false OK
ok 17 - pclose2 failed with EIO
1..17
```

garlick commented:
> How is that going to work for flux-core installed in /opt? Should ssh connector exec a fully qualified path to flux, determined from the calling side by default but overridden by an environment variable?

That doesn't seem clear cut.  Will people just load the flux-core module in their dot files?  That would take care of it.

grondo commented:
> Will people just load the flux-core module in their dot files? That would take care of it.

Does non-interactive rsh run any rc files before executing its command?
I'm actually having a bit of trouble resetting `TMPDIR` for the `rsh flux proxy` command. `.bashrc` doesn't work for me...

grondo commented:
Anyhow, my opinion is that there should at least be an environment variable to set an explicit path to `flux(1)` for the ssh connector. Then I can actually run tests against my build directory without having to `make install` it somehwere and modify all my rc-files. The nice thing about `flux(1)` is it handles all the environment encapsulation nicely once you run it.

garlick commented:
Hmm, it worked for me:
```
$ rsh hype346 printenv TMPDIR
/var/tmp/garlick
$ cat .bashrc
PATH=$HOME/local/bin:$PATH
export LD_LIBRARY_PATH=$HOME/local/lib:$PATH
export TMPDIR=/var/tmp/garlick
```

grondo commented:
Ok, I had one of these age old things in my `.bashrc`
```bash
# If not running interactively, don't do anything:
[ -z "$PS1" ] && return
```

garlick commented:
> here should at least be an environment variable to set an explicit path to flux(1) for the ssh connector.

That sounds reasonable.  Maybe FLUX_SSH_REMOTEFLUX?

grondo commented:
Ok, after getting TMPDIR set appropriately for the remote command (either /tmp/grondo or /var/tmp/grondo would work). I can directly run `flux proxy` on hype login node:
```
$ src/cmd/flux proxy --uri=ssh://hype347?jobid=1147793.0
(flux-proxy-) grondo@hype356:~/git/f $ 
```

However, `flux` commands seem to just hang, messages go out but not back in.

grondo commented:
hm, maybe more subtle than that, but this seems really close to working! Maybe some buffering issue with mrsh?

```
 grondo@hype356:~/git/f$ src/cmd/flux -t proxy --uri=ssh://hype347?jobid=1147793.0
--------------------------------------
>[000] ||
>[011] cmb.attrget
>[018] { "name": "rank" }
>[012] 8E01010FFFFFFFFF00000001
--------------------------------------
<[000] ||
<[011] cmb.attrget
<[028] { "value": "0", "flags": 5 }
<[012] 8E01020F0000000000000001
(flux-proxy-) grondo@hype356:~/git/f$ flux up
--------------------------------------
>[000] ||
>[007] kvs.get
>[028] { "conf.live.status": null }
>[012] 8E01010FFFFFFFFF00000001
--------------------------------------
>[032] |15D90|
>[007] kvs.get
>[028] { "conf.live.status": null }
>[012] 8E01010FFFFFFFFF00000001

[hangs]
```

garlick commented:
Yep that is what I am seeing.  Given that the tests are passing I suspect
something to do with the way data is being chunked by the SSH tunnel
exposing some bug which we don't see when the components all talk directly
with each other as they do in test.
On Apr 20, 2016 11:23 AM, "Mark Grondona" <notifications@github.com> wrote:

> Ok, after getting TMPDIR set appropriately for the remote command (either
> /tmp/grondo or /var/tmp/grondo would work). I can directly run flux proxy
> on hype login node:
>
> $ src/cmd/flux proxy --uri=ssh://hype347?jobid=1147793.0
> (flux-proxy-) grondo@hype356:~/git/f $
>
> However, flux commands seem to just hang, messages go out but not back in.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/648#issuecomment-212544397>
>

garlick commented:
To clarify - I see the same behavior with mrsh and ssh, but I don't see it with my "scripts/tssh" program that just runs stuff under a shell (after gobbling the ssh arguments) for t/t1008-proxy.t.

grondo commented:
FYI, OpenSSH 6.7 will supposedly support Unix domain socket forwarding [directly](http://lwn.net/Articles/609321/). Unfortunately looks like even our RHEL7 systems only have 6.6

garlick commented:
> OpenSSH 6.7

Interesting!

OK, I'm packing up and will be offline for a while.  Will have another go
at it this evening.


On Wed, Apr 20, 2016 at 1:04 PM, Mark Grondona <notifications@github.com>
wrote:

> FYI, OpenSSH 6.7 will supposedly support Unix domain socket forwarding
> directly <http://lwn.net/Articles/609321/>. Unfortunately looks like even
> our RHEL7 systems only have 6.6
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/648#issuecomment-212581945>
>

grondo commented:
For fun, I tried to see if `socat` would work at all.

It doesn't have any job-finding logic, so in a running instance, check the URI
```
$ echo $FLUX_URI
local:///var/tmp/grondo/flux-1147793.0-wWGRNm/0
```

On the local host then:
```
$ mkdir /tmp/grondo/test/0
$ echo $$ >/tmp/grondo/test/0/broker.pid
$ /path/to/socat "UNIX-LISTEN:/tmp/grondo/test/0/local,reuseaddr,fork" \
    EXEC:"rsh <remotehost> /path/to/socat STDIO UNIX-CONNECT\:/tmp/grondo/flux-1147793.0-wWGRNm/0/local"
```

In another window on the same machine:
```
$ FLUX_URI=local:///tmp/grondo/test/0 src/cmd/flux env bash
grondo@hype356:~/git/f$ flux up
ok:     [0-3]
slow:   
fail:   
unknown:
grondo@hype356:~/git/f$ flux kvs dir
resource.
lwj-active.
lwj.

```

grondo commented:
Of course I should have used `flux proxy` instead of `flux env bash`:
```
$  src/cmd/flux proxy --uri=local:///tmp/grondo/test/0 
(flux-proxy) grondo@hype356:~/git/f$ flux up
ok:     [0-3]
slow:   
fail:   
unknown:
grondo@hype356:~/git/f$ flux wreckrun -n4 hostname
hype348
hype349
hype350
hype347
```

`flux proxy` definitely works!

garlick commented:
Thanks @grondo.  I am offline today but I will get back on this Fri.
On Apr 20, 2016 1:56 PM, "Mark Grondona" <notifications@github.com> wrote:

> Of course I should have used flux proxy instead of flux env bash:
>
> $  src/cmd/flux proxy --uri=local:///tmp/grondo/test/0
> (flux-proxy) grondo@hype356:~/git/f$ flux up
> ok:     [0-3]
> slow:
> fail:
> unknown:
> grondo@hype356:~/git/f$ flux wreckrun -n4 hostname
> hype348
> hype349
> hype350
> hype347
>
> flux proxy definitely works!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/648#issuecomment-212603295>
>

garlick commented:
Just rebased on current master and squashed incremental development.

garlick commented:
This last commit seems to have fixed the outstanding problem with the ssh:// connector.    I don't completely understand why the fix works yet and need to head home.  Will update here once I figure it out.

garlick commented:
Small progress: added a test that fails before 668c9e1 and succeeds after.  Real ssh is not required to make it fail, it turns out.  `flux proxy --uri ssh://...` will do it with `tssh` which is just a shell that directly runs the "remote command", which is `flux proxy --stdio`.

I also implemented @grondo's suggestion of an env var (FLUX_SSH_RCMD) to set the path to remote flux command.

garlick commented:
Ah, found the issue.  The ssh:// connector does a `flux_attr_get()` during initialization so that fatal errors in the remote command can be detected inside `flux_open()`. This triggers the creation of a dispatcher, which in turn calls `flux_get_reactor()`, implicitly creating a reactor.

`flux-proxy` then creates its own reactor and calls `flux_set_reactor()`.  Although this replaces the reactor so that `flux_get_reactor()` will return the new one, it doesn't tell the dispatcher which has created a handle watcher on the first one.  So when `flux-proxy` runs its reactor, the dispatcher is not wired to it.

So I would say there are two problems here.  One is that the ssh:// connector should not call `flux_attr_get()` but should instead use low level calls to avoid triggering dispatcher or reactor setup.  Two is that `flux_set_reactor()` should probably return an error if the reactor has already been created instead of just dropping a refcount and replacing it, and the documentation should be made clear that `flux_set_reactor()` can only be called immediately after `flux_open()`.

garlick commented:
Just forced a push with a targeted fix, as well as the other collateral fixes.

With that diversion out of the way, here's a recap of review comments and status:

Addressed:
* `pxopen()` was renamed to `popen2()`
* Usage is now `flux proxy --uri URI`, not `flux --uri URI proxy` or `flux proxy`
* Added env variable for setting remote `flux` path (FLUX_RSH_RCMD)

Pending:
* FLUX_SSH should default to ssh not rsh
* replace `--jobid` with a URI scheme for jobs

Other todo:
* Retest on hype
* Verify that FLUX_RSH_RCMD works on hype (LD_LIBRARY_PATH may be a problem)
* Fix intermittent `popen2()` test failure, if seen again

garlick commented:
> Verify that FLUX_RSH_RCMD works on hype (LD_LIBRARY_PATH may be a problem)

I tried this with a git tree on both sides.  Started a size=8 job no hype login node, then on desktop ran
```
FLUX_SSH_RCMD=/g/g0/garlick/flux-core/src/cmd/flux \
  ./flux proxy --uri ssh://hype?jobid=130928
```
I had to have TMPDIR set in my .bashrc on hype for the remote proxy to find my job.

This also works for flux "side installed" to $HOME/local, so I presume it will work for /opt once `flux-proxy` is available there.

It was not necessary to set LD_LIBRARY_PATH in either case.

grondo commented:
> Usage is now flux proxy --uri URI, not flux --uri URI proxy or flux proxy

I also tried to argue for `flux proxy URI` and drop both `--uri` and `--jobid`. Jobid can be detected via specialized URI (e.g. `job://ID`) or by heuristic that assumes if argument doesn't look like a URI then it is a job identifier.

I argue for this for the following reasons
 * "Required" options generally make an awkward UI
 * It is more difficult to check for at least one and only one option during option processing
 * the uri scheme is more easily extensible, rather than adding more options

I don't know if you saw this argument already and dismissed it, sorry if so!
I admit I could be wrong here and it is a minor issue.

garlick commented:
I tried to capture that above in
> replace --jobid with a URI scheme for jobs

I'm having a hard time getting my head around this because I viewed these URI schemes as fundamentally selecting a transport.  E.g. providing a connector plugin so job://  can be passed to `flux_open()` seems pretty odd to me, especially since the connector is somewhat limited in what it can do (as evidenced by bug I just fixed :-).

At the moment you can say `flux proxy --uri ssh://hype&jobid=42 flux up` or equivalently (with less plumbing) `flux --uri ssh://hype&jobid=42 up`, since `flux-proxy` just passes its uri argument to `flux_open()`.  Were you envisioning job:// parsed just in `flux-proxy`, not passed to `flux-open()`?

Sorry I feel I'm being thick headed here :-/

grondo commented:
Yes, flux-proxy should process its argument before passing it to `flux_open()`  :smile:

I guess for user-facing commands I would opt to simplify the interface and complicate the implementation a bit. Having the `flux proxy --uri | --jobid` syntax for me is like nails on chalkboard.
It would be like `wget` or `curl` requiring a `--url` option that you always have to use `--url=http://` or ssh requiring `--host` or `--ip` options.

It would be sufficient if the usage was something like `flux proxy URI` and `flux proxy --jobid JOBID` where `--jobid` is a switch stating that the argument to `flux proxy` is a jobid not a "connector URI", though I think in the end someone would eventually submit a patch to autodetect jobid if `flux proxy` was regularly used

Sorry I think I'm actually making more out of this than is really there, I'm willing to drop it.

garlick commented:
Oh duh, I get it.  How about:

```
flux proxy JOB [CMD ...]
```
where JOB is either a URI passed to `flux_open()` or if it lacks `://` substring, some other heuristically parsed job identifier that can be used to find a URI.  Maybe even slurm:jobid?

Sorry to make you work so hard!

grondo commented:
> How about:
> `flux proxy JOB [CMD ...]`

Perfect! Something like that is exactly what I was thinking!

garlick commented:
Well that simplified the code a bit.  ssh:// now just blindly passes the "path" component as the JOB argument to `flux --stdio JOB`, so `ssh://hype/42` just works now.

Before that change, I had added a `--setenv NAME=VAR` option to `flux-proxy` so you can also say `ssh://hype/42?setenv=TMPDIR=/var/tmp/garlick`.

This seems better than before, thanks @grondo for the good suggestions (and for being persistent).

garlick commented:
This was rebased on current master and incremental development squashed.  Maybe ready?

garlick commented:
Perhaps silly of me to ask at this late stage, but would `flux-attach` be a better name than `flux-proxy`?

grondo commented:
I don't know why but I still like flux-proxy.

So this one is pretty much ready? I'll try to poke at it tonight but it all looks good to me

garlick commented:
OK, fine with me.  Yes I'd be OK with merging this now.

grondo commented:
Ok, I poked at this a bit and it is looking great. Perhaps documentation for `FLUX_SSH_RCMD` should be added in `flux-proxy(1)` man page since there is some documentation of `ssh` connector there too? (I had to look at source code to remember what it was)

I think an (perhaps weak) argument could still be made that `flux proxy` should automatically set `FLUX_SSH_RCMD` to the absolute path to the currently executing `flux`, unless overridden by existing `FLUX_SSH_RCMD`. For instance, this would make `flux proxy` "just work" for side-installs, the `/opt` version, and working directory invocations. Otherwise, in all three of these cases, users are required to code `FLUX_SSH_RCMD` appropriately by hand.

Also, is this of any concern? From local connector:
``` 
$ time (flux kvs dir -r >/dev/null)
real	0m0.973s
user	0m0.166s
sys	0m0.067s
```

Under `flux-proxy`:
```
$ time (flux kvs dir -r >/dev/null)

real	5m10.774s
user	0m0.212s
sys	0m0.326s
```

Almost 300x slowdown? Am I doing something wrong or is this a particularly bad case due to no local content cache with `flux-proxy`. I was running `flux-proxy` on hype login node to a job running on hype302 via
```
FLUX_SSH_RCMD=$(pwd)/src/cmd/flux src/cmd/flux proxy ssh://hype302/1148889.2
```

garlick commented:
Hmm, that's interesting.  Here's a couple more data points using a job launched locally

Hype login node, direct proxy versus proxy via mrsh:
```
$ time ./flux proxy 174044 flux kvs dir -r >/dev/null

real	0m0.967s
user	0m0.313s
sys	0m0.207s

$ time FLUX_SSH_RCMD=$(pwd)/flux \
  ./flux proxy ssh://localhost/174044 flux kvs dir -r >/dev/null

real	3m19.577s
user	0m0.404s
sys	0m1.000s
```

Same experiment on my desktop with ssh instead of mrsh:
```
$ time ./flux proxy 48414 flux kvs dir -r  >/dev/null

real	0m0.703s
user	0m0.200s
sys	0m0.164s

$ time FLUX_SSH_RCMD=$(pwd)/flux \
  ./flux proxy ssh://localhost/48414 flux kvs dir -r >/dev/null

real	0m1.869s
user	0m0.380s
sys	0m0.328s
```

garlick commented:
Same test on hype but using ssh to localhost (moved time so that interactive 2-factor auth is not included)
```
$ FLUX_SSH_RCMD=$(pwd)/flux FLUX_SSH=/usr/bin/ssh \
  ./flux proxy ssh://localhost/174044 time flux kvs dir -r >/dev/null

real	0m1.306s
user	0m0.148s
sys	0m0.053s
```
so it seems like mrsh is implicated but I have no idea why.

garlick commented:
Interesting the message changes I just pushed seem to have fixed the problem.  The middle test is the one that uses mrsh that took 3m19s above.
```
$ ./flux proxy 139943 time flux kvs dir -r >/dev/null

real	0m0.914s
user	0m0.151s
sys	0m0.054s

$ FLUX_SSH_RCMD=$(pwd)/flux \
  ./flux proxy ssh://localhost/139943 time flux kvs dir -r >/dev/null

real	0m1.170s
user	0m0.128s
sys	0m0.056s

$ FLUX_SSH=/usr/bin/ssh FLUX_SSH_RCMD=$(pwd)/flux \
  ./flux proxy ssh://localhost/139943 time flux kvs dir -r >/dev/null
[snip password promt etc]
real	0m1.367s
user	0m0.135s
sys	0m0.057s
```

grondo commented:
Oh, nice work!

garlick commented:
I didn't really expect that result.  @chu11 verified that [mrsh](http://github.com/chaos/mrsh) doesn't set the TCP_NODELAY socket option, so TCP Nagle _should_ have been masking the old code's small writes.   I was heading down the path of trying to disable Nagle in mrsh to see if TCP buffering delays were increasing RPC latency.

I wish I understood what is happening... But apparently less stupid code goes faster.  Maybe I should accept that bit of wisdom and move on :-)

garlick commented:
> flux proxy should automatically set FLUX_SSH_RCMD to the absolute path to the currently executing flux, unless overridden by existing FLUX_SSH_RCMD

I agree.  I'm just puzzling over how to obtain that from within the connector.  Did you have some thoughts?

grondo commented:
>  I agree. I'm just puzzling over how to obtain that from within the connector. Did you have some thoughts?

Since the main use case for now is `flux-proxy`, could `flux-proxy` itself just set `FLUX_SSH_RCMD` appropriately unless overridden in current environment? If that would be considered surprising to some users (e.g. if running `flux proxy` from a side install on a desktop to a cluster where you *want* to use flux in your PATH), I'd settle for an option to do this magic.

garlick commented:
Just pushed the FLUX_SSH_RCMD fix - I just searched for `flux` on the PATH in the `ssh://` connector.

If the recent changes are agreeable I'll sqaush down the incremental development and then maybe this is ready?

grondo commented:
Yeah, looks good to me! I was thinking ssh connector should sanity check `FLUX_SSH_RCMD` since it will be exec-ing it directly, e.g. if it is set but empty, that might produce an undesireable result. However, I don't think that it is all that important.

garlick commented:
Oh I can throw that in real quick.

garlick commented:
OK just forced a push with incremental work squashed and the check for empty FLUX_SSH_RCMD.

grondo commented:
Great! Merging!
